File: partiql-ast/src/main/java/org/partiql/ast/v1/SelectProject.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public class SelectProject extends Select {
    @NotNull
    public List<ProjectItem> items;

    @Nullable
    public SetQuantifier setq;

    public SelectProject(@NotNull List<ProjectItem> items, @Nullable SetQuantifier setq) {
        this.items = items;
        this.setq = setq;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        return new ArrayList<>(items);
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitSelectProject(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/SelectProject.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class SelectProject(
    @JvmField
    public var items: List<ProjectItem>,
    @JvmField
    public var setq: SetQuantifier?,
) : Select() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.addAll(items)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitSelectProject(this, ctx)
}
